ENH: pickle serialization for itk images (issue 1091)

GenevieveBuckley opened the pull request:
This PR adds pickle support to ITK images. We want this so that ITK images can be serialized for use with Dask arrays.

Closes https://github.com/InsightSoftwareConsortium/ITK/issues/1091

cc @thewtex 

<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start ITK commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

<!-- **Thanks for contributing to ITK!** -->

GenevieveBuckley commented:
This is not going to pass the CI checks right now. There's a few more steps:

- [ ] 1. We need to add the functions `itkimage_to_json` & `itkimage_to_json` (plus their helper functions `_image_to_type` & `_type_to_image`). **Question: Where do these need to live?** 

<details>
<summary>Click to expand details to see what python code we need to add:</summary>

```python
import os
import six
from functools import reduce
import numpy as np
try:
    import zstandard as zstd
except ImportError:
    import zstd

import itk


def _image_to_type(itkimage):  # noqa: C901
    component = itk.template(itkimage)[1][0]
    if component == itk.UL:
        if os.name == 'nt':
            return 'uint32_t', 1
        else:
            return 'uint64_t', 1
    mangle = None
    pixelType = 1
    if component == itk.SL:
        if os.name == 'nt':
            return 'int32_t', 1,
        else:
            return 'int64_t', 1,
    if component in (itk.SC, itk.UC, itk.SS, itk.US, itk.SI, itk.UI, itk.F,
            itk.D, itk.B):
        mangle = component
    elif component in [i[1] for i in itk.Vector.items()]:
        mangle = itk.template(component)[1][0]
        pixelType = 5
    elif component == itk.complex[itk.F]:
        # complex float
        return 'float', 10
    elif component == itk.complex[itk.D]:
        # complex float
        return 'double', 10
    elif component in [i[1] for i in itk.CovariantVector.items()]:
        # CovariantVector
        mangle = itk.template(component)[1][0]
        pixelType = 7
    elif component in [i[1] for i in itk.Offset.items()]:
        # Offset
        return 'int64_t', 4
    elif component in [i[1] for i in itk.FixedArray.items()]:
        # FixedArray
        mangle = itk.template(component)[1][0]
        pixelType = 11
    elif component in [i[1] for i in itk.RGBAPixel.items()]:
        # RGBA
        mangle = itk.template(component)[1][0]
        pixelType = 3
    elif component in [i[1] for i in itk.RGBPixel.items()]:
        # RGB
        mangle = itk.template(component)[1][0]
        pixelType = 2
    elif component in [i[1] for i in itk.SymmetricSecondRankTensor.items()]:
        # SymmetricSecondRankTensor
        mangle = itk.template(component)[1][0]
        pixelType = 8
    else:
        raise RuntimeError('Unrecognized component type: {0}'.format(str(component)))
    _python_to_js = {
        itk.SC: 'int8_t',
        itk.UC: 'uint8_t',
        itk.SS: 'int16_t',
        itk.US: 'uint16_t',
        itk.SI: 'int32_t',
        itk.UI: 'uint32_t',
        itk.F: 'float',
        itk.D: 'double',
        itk.B: 'uint8_t'
    }
    return _python_to_js[mangle], pixelType


def itkimage_to_json(itkimage, manager=None):
    """Serialize a Python itk.Image object.
    Attributes of this dictionary are to be passed to the JavaScript itkimage
    constructor.
    """
    if itkimage is None:
        return None
    else:
        direction = itkimage.GetDirection()
        directionMatrix = direction.GetVnlMatrix()
        directionList = []
        dimension = itkimage.GetImageDimension()
        pixel_arr = itk.array_view_from_image(itkimage)
        componentType, pixelType = _image_to_type(itkimage)
        if 'int64' in componentType:
            # JavaScript does not yet support 64-bit integers well
            if componentType == 'uint64_t':
                pixel_arr = pixel_arr.astype(np.uint32)
                componentType = 'uint32_t'
            else:
                pixel_arr = pixel_arr.astype(np.int32)
                componentType = 'int32_t'
        compressor = zstd.ZstdCompressor(level=3)
        compressed = compressor.compress(pixel_arr.data)
        pixel_arr_compressed = memoryview(compressed)
        for col in range(dimension):
            for row in range(dimension):
                directionList.append(directionMatrix.get(row, col))
        imageType = dict(
            dimension=dimension,
            componentType=componentType,
            pixelType=pixelType,
            components=itkimage.GetNumberOfComponentsPerPixel()
        )
        return dict(
            imageType=imageType,
            origin=tuple(itkimage.GetOrigin()),
            spacing=tuple(itkimage.GetSpacing()),
            size=tuple(itkimage.GetBufferedRegion().GetSize()),
            direction={'data': directionList,
                       'rows': dimension,
                       'columns': dimension},
            compressedData=compressed
        )


def _type_to_image(jstype):
    _pixelType_to_prefix = {
        1: '',
        2: 'RGB',
        3: 'RGBA',
        4: 'O',
        5: 'V',
        7: 'CV',
        8: 'SSRT',
        11: 'FA'
    }
    pixelType = jstype['pixelType']
    dimension = jstype['dimension']
    if pixelType == 10:
        if jstype['componentType'] == 'float':
            return itk.Image[itk.complex, itk.F], np.float32
        else:
            return itk.Image[itk.complex, itk.D], np.float64

    def _long_type():
        if os.name == 'nt':
            return 'LL'
        else:
            return 'L'
    prefix = _pixelType_to_prefix[pixelType]
    _js_to_python = {
        'int8_t': 'SC',
        'uint8_t': 'UC',
        'int16_t': 'SS',
        'uint16_t': 'US',
        'int32_t': 'SI',
        'uint32_t': 'UI',
        'int64_t': 'S' + _long_type(),
        'uint64_t': 'U' + _long_type(),
        'float': 'F',
        'double': 'D'
    }
    _js_to_numpy_dtype = {
        'int8_t': np.int8,
        'uint8_t': np.uint8,
        'int16_t': np.int16,
        'uint16_t': np.uint16,
        'int32_t': np.int32,
        'uint32_t': np.uint32,
        'int64_t': np.int64,
        'uint64_t': np.uint64,
        'float': np.float32,
        'double': np.float64
    }
    dtype = _js_to_numpy_dtype[jstype['componentType']]
    if pixelType != 4:
        prefix += _js_to_python[jstype['componentType']]
    if pixelType not in (1, 2, 3, 10):
        prefix += str(dimension)
    prefix += str(dimension)
    return getattr(itk.Image, prefix), dtype


def itkimage_from_json(js, manager=None):
    """Deserialize a Javascript itk.js Image object."""
    if js is None:
        return None
    else:
        ImageType, dtype = _type_to_image(js['imageType'])
        decompressor = zstd.ZstdDecompressor()
        if six.PY2:
            asBytes = js['compressedData'].tobytes()
            pixelBufferArrayCompressed = np.frombuffer(asBytes, dtype=np.uint8)
        else:
            pixelBufferArrayCompressed = np.frombuffer(js['compressedData'],
                                                    dtype=np.uint8)
        pixelCount = reduce(lambda x, y: x * y, js['size'], 1)
        numberOfBytes = pixelCount * \
            js['imageType']['components'] * np.dtype(dtype).itemsize
        pixelBufferArray = \
            np.frombuffer(decompressor.decompress(pixelBufferArrayCompressed,
                                                numberOfBytes),
                        dtype=dtype)
        pixelBufferArray.shape = js['size'][::-1]
        # Workaround for GetImageFromArray required until 5.0.1
        # and https://github.com/numpy/numpy/pull/11739
        pixelBufferArrayCopyToBeRemoved = pixelBufferArray.copy()
        # image = itk.PyBuffer[ImageType].GetImageFromArray(pixelBufferArray)
        image = itk.PyBuffer[ImageType].GetImageFromArray(
            pixelBufferArrayCopyToBeRemoved)
        Dimension = image.GetImageDimension()
        image.SetOrigin(js['origin'])
        image.SetSpacing(js['spacing'])
        direction = image.GetDirection()
        directionMatrix = direction.GetVnlMatrix()
        directionJs = js['direction']['data']
        for col in range(Dimension):
            for row in range(Dimension):
                directionMatrix.put(
                    row, col, directionJs[col + row * Dimension])
        return image

```

</details>

- [ ] 2. Add a test like `pickle.loads(pickle.dumps(itk_image))` **Question: Where should the test live?** 

I see a "Testing" folder in the ITK repo, but I haven't narrowed it down any further. Also, I don't see tests written in the python language, so I'm not totally sure how we'd go about testing a python specific thing like pickle.

GenevieveBuckley commented:
I'm a little bit worried I don't see the tests FAIL at commit eb18f29. They should be failing here, because the code is still incomplete and I've added this test:
```
# Test serialization with pickle
assert (pickle.loads(pickle.dumps(image)) == image).all()
```

Thoughts:
* Maybe image is not the right kind of ITK image?
* Maybe the CI isn't running that particular check (seems super unlikely)

I'm going to go ahead and push the rest of my commits, but I'm hoping these links to the CI builds will let you see the build for this moment in time, when we expect to see test failures and don't:
- [ITK.Linux](https://dev.azure.com/itkrobotlinux/ITK.Linux/_build/results?buildId=6821&view=results)
- [ITK.Linux.Python](https://dev.azure.com/itkrobotlinuxpython/ITK.Linux.Python/_build/results?buildId=5453&view=results)
- [ITK.Windows](https://dev.azure.com/itkrobotwindow/ITK.Windows/_build/results?buildId=6796&view=results)
- [ITK.macOSPython](https://dev.azure.com/itkrobotmacospython/ITK.macOS.Python/_build/results?buildId=5503&view=logs&j=a5e52b91-c83f-5429-4a68-c246fc63a4f7&t=5add3200-7efa-59c0-0a45-43bce0c77c69)
- [circleci build](https://circleci.com/gh/InsightSoftwareConsortium/ITK/10437?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

GenevieveBuckley commented:
Things to check in the review:
1. Are my imports correct (I think they are, but I don't have a local ITK dev env setup, so I can't check this easily)
2. Do the tests behave as expected (locally & on CI)

thewtex commented on Wrapping/Generators/Python/itk/support/extras.py:
```diff
@@ -801,6 +806,59 @@ def image_from_vtk_image(vtk_image: "vtk.vtkImageData") -> "itkt.ImageBase":
+    import itk
+
+    ImageType, dtype = type_to_image(image_dict['imageType'])
+    image = itk.PyBuffer[ImageType].GetImageFromArray(image_dict['data'])
```
Yes, we want / need to roughtly do it this way, but `GetImageFromArray` -> `GetImageViewFromArray`

thewtex commented on Wrapping/Generators/Python/itk/support/extras.py:
```diff
@@ -801,6 +806,53 @@ def image_from_vtk_image(vtk_image: "vtk.vtkImageData") -> "itkt.ImageBase":
+    image.SetOrigin(image_dict['origin'])
+    image.SetSpacing(image_dict['spacing'])
+    direction = image_dict['direction']['data']
+    image.SetDirection(direction)
```
> self.__dict__['this'] = deserialized

That's cool! If it works, then, yes, maybe that is all we need.

thewtex commented on Wrapping/Generators/Python/PyBase/pyBase.i:
```diff
@@ -522,6 +522,22 @@ str = str
+
+            def __setstate__(self, state):
+                """Set object state, necessary for serialization with pickle."""
+                import itk
```
import numpy as np

thewtex reviewed (commented):
>I think I've done these two things now. We'll see what the CI tests say.

Awesome, thanks, Genevieve!

>I have not moved zstandard compression into __getstate__/__setstate__, right now it's been completely removed. It's my understanding that you want to try this out without any compression, is that correct @thewtex?

Yes, we could possibly add the compression in a subsequent step. For distributed computing with Dask distributed, the compute / data transfer may make it worthwhile (?)

Regarding the size / typing errors, we do need to use:

```
    ImageType = helpers.image_type_from_wasm_type(image_dict['imageType'])
    image = itk.PyBuffer[ImageType].GetImageViewFromArray(image_dict['data'])
```

Where `type_to_image` is renamed to `image_type_from_wasm_type` and it only returns the `ImageType`.

The reason for this is that a `np.ndarray` does not contain all the typing information as an `imageType`. The *dtype* corresponds to the itk *componentType*. However, the *imageType* also defines how many spatial dimensions there are, if it is multi-component image, how many components there are and what those components mean, i.e. are they RGB values or values of a co-variant vector or values of a contravariant vector, etc.

Additionally, ideally we want this serialization format to correspond to what is used for interfacing with WebAssembly, described here:

- https://github.com/InsightSoftwareConsortium/itk-wasm/blob/master/doc/content/api/ImageType.md
- https://github.com/InsightSoftwareConsortium/itk-wasm/blob/master/doc/content/api/Image.md

GenevieveBuckley commented:
Have tried to address review comments, waiting on CI checks now.

GenevieveBuckley commented:
ITK doesn't seem to recognize the array type, I'm getting messages to say it's not wrapped. We're using a `uint8` test array. I've just adjusted the test to try the default integer type (`int64`) just to see what happens.

The recommended solution is to set `ITK_WRAP_${type}`. I'm not entirely sure:
1. Where to find this in the code - searching for "`ITK_WRAP_`" across the repository returns 4,800 results
2. I'm not sure how many extra different types we'd need to wrap. When I run `itk.PyBuffer.GetTypes()` it says that `itkImageUC2` **is** in the list. So I have to assume that it's the extra `uint8` bit at the end that is causing the difference and also needs wrapping? If that's the case, there might be many other array dtypes that also need the same treatment (...but also, the ITK message says that we don't want to just wrap everything, because then that makes the package too big. So I don't know how to choose what's needed.)

```python-traceback
2021-11-01T04:22:29.5169190Z 150: Traceback (most recent call last):
2021-11-01T04:22:29.5245040Z 150:   File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/support/template_class.py", line 525, in __getitem__
2021-11-01T04:22:29.5276550Z 150:     this_item = self.__template__[key]
2021-11-01T04:22:29.5287930Z 150: KeyError: (<class 'itk.itkImagePython.itkImageUC2'>, <class 'numpy.uint8'>)
2021-11-01T04:22:29.5293920Z 150: 
2021-11-01T04:22:29.5349140Z 150: During handling of the above exception, another exception occurred:
2021-11-01T04:22:29.5378830Z 150: 
2021-11-01T04:22:29.5451240Z 150: Traceback (most recent call last):
2021-11-01T04:22:29.5480370Z 150:   File "/Users/runner/work/1/s/Wrapping/Generators/Python/Tests/extras.py", line 176, in <module>
2021-11-01T04:22:29.5552620Z 150:     assert (pickle.loads(pickle.dumps(image)) == image).all()
2021-11-01T04:22:29.5582710Z 150:   File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/itkImagePython.py", line 4451, in __setstate__
2021-11-01T04:22:29.5654270Z 150:     deserialized = itk.image_from_dict(state)
2021-11-01T04:22:29.5684970Z 150:   File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/support/extras.py", line 844, in image_from_dict
2021-11-01T04:22:29.5754860Z 150:     image = itk.PyBuffer[ImageType].GetImageViewFromArray(image_dict['data'])
2021-11-01T04:22:29.5787390Z 150:   File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/support/template_class.py", line 529, in __getitem__
2021-11-01T04:22:29.5858280Z 150:     raise itk.TemplateTypeError(self, key)
2021-11-01T04:22:29.5890140Z 150: itk.support.extras.TemplateTypeError: itk.PyBuffer is not wrapped for input type `itk.Image[itk.UC,2], uint8`.
```

<details>
<summary>Full details - click to expand!:</summary>

```python-traceback
2021-11-01T04:22:29.5169190Z 150: Traceback (most recent call last):
2021-11-01T04:22:29.5245040Z 150:   File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/support/template_class.py", line 525, in __getitem__
2021-11-01T04:22:29.5276550Z 150:     this_item = self.__template__[key]
2021-11-01T04:22:29.5287930Z 150: KeyError: (<class 'itk.itkImagePython.itkImageUC2'>, <class 'numpy.uint8'>)
2021-11-01T04:22:29.5293920Z 150: 
2021-11-01T04:22:29.5349140Z 150: During handling of the above exception, another exception occurred:
2021-11-01T04:22:29.5378830Z 150: 
2021-11-01T04:22:29.5451240Z 150: Traceback (most recent call last):
2021-11-01T04:22:29.5480370Z 150:   File "/Users/runner/work/1/s/Wrapping/Generators/Python/Tests/extras.py", line 176, in <module>
2021-11-01T04:22:29.5552620Z 150:     assert (pickle.loads(pickle.dumps(image)) == image).all()
2021-11-01T04:22:29.5582710Z 150:   File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/itkImagePython.py", line 4451, in __setstate__
2021-11-01T04:22:29.5654270Z 150:     deserialized = itk.image_from_dict(state)
2021-11-01T04:22:29.5684970Z 150:   File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/support/extras.py", line 844, in image_from_dict
2021-11-01T04:22:29.5754860Z 150:     image = itk.PyBuffer[ImageType].GetImageViewFromArray(image_dict['data'])
2021-11-01T04:22:29.5787390Z 150:   File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/support/template_class.py", line 529, in __getitem__
2021-11-01T04:22:29.5858280Z 150:     raise itk.TemplateTypeError(self, key)
2021-11-01T04:22:29.5890140Z 150: itk.support.extras.TemplateTypeError: itk.PyBuffer is not wrapped for input type `itk.Image[itk.UC,2], uint8`.
2021-11-01T04:22:29.5962970Z 150: 
2021-11-01T04:22:29.5995970Z 150: To limit the size of the package, only a limited number of
2021-11-01T04:22:29.6064550Z 150: types are available in ITK Python. To print the supported
2021-11-01T04:22:29.6097690Z 150: types, run the following command in your python environment:
2021-11-01T04:22:29.6113420Z 150: 
2021-11-01T04:22:29.6171850Z 150:     itk.PyBuffer.GetTypes()
2021-11-01T04:22:29.6214970Z 150: 
2021-11-01T04:22:29.6273310Z 150: Possible solutions:
2021-11-01T04:22:29.6299720Z 150: * If you are an application user:
2021-11-01T04:22:29.6317900Z 150: ** Convert your input image into a supported format (see below).
2021-11-01T04:22:29.6376120Z 150: ** Contact developer to report the issue.
2021-11-01T04:22:29.6402390Z 150: * If you are an application developer, force input images to be
2021-11-01T04:22:29.6419710Z 150: loaded in a supported pixel type.
2021-11-01T04:22:29.6478930Z 150: 
2021-11-01T04:22:29.6504980Z 150:     e.g.: instance = itk.PyBuffer[itk.Image[itk.SS,2]].New(my_input)
2021-11-01T04:22:29.6521580Z 150: 
2021-11-01T04:22:29.6581470Z 150: * (Advanced) If you are an application developer, build ITK Python yourself and
2021-11-01T04:22:29.6606990Z 150: turned to `ON` the corresponding CMake option to wrap the pixel type or image
2021-11-01T04:22:29.6623300Z 150: dimension you need. When configuring ITK with CMake, you can set
2021-11-01T04:22:29.6683270Z 150: `ITK_WRAP_${type}` (replace ${type} with appropriate pixel type such as
2021-11-01T04:22:29.6708610Z 150: `double`). If you need to support images with 4 or 5 dimensions, you can add
2021-11-01T04:22:29.6724810Z 150: these dimensions to the list of dimensions in the CMake variable
2021-11-01T04:22:29.6784960Z 150: `ITK_WRAP_IMAGE_DIMS`.
2021-11-01T04:22:29.6810720Z 150: 
2021-11-01T04:22:29.6824150Z 150: Supported input types:
2021-11-01T04:22:29.6866850Z 150: 
2021-11-01T04:22:29.6886940Z 150: itk.Image[itk.SS,2]
2021-11-01T04:22:29.6928350Z 150: itk.Image[itk.SS,3]
2021-11-01T04:22:29.6968720Z 150: itk.Image[itk.UC,2]
2021-11-01T04:22:29.6988820Z 150: itk.Image[itk.RGBPixel[itk.UC],2]
2021-11-01T04:22:29.7031540Z 150: itk.Image[itk.RGBAPixel[itk.UC],2]
2021-11-01T04:22:29.7070310Z 150: itk.Image[itk.UC,3]
2021-11-01T04:22:29.7090230Z 150: itk.Image[itk.RGBPixel[itk.UC],3]
2021-11-01T04:22:29.7128190Z 150: itk.Image[itk.RGBAPixel[itk.UC],3]
2021-11-01T04:22:29.7132950Z 150: itk.Image[itk.F,2]
2021-11-01T04:22:29.7193510Z 150: itk.Image[itk.SymmetricSecondRankTensor[itk.F,2],2]
2021-11-01T04:22:29.7229970Z 150: itk.Image[itk.Vector[itk.F,2],2]
2021-11-01T04:22:29.7236270Z 150: itk.Image[itk.CovariantVector[itk.F,2],2]
2021-11-01T04:22:29.7295040Z 150: itk.Image[itk.Vector[itk.F,3],2]
2021-11-01T04:22:29.7334710Z 150: itk.Image[itk.CovariantVector[itk.F,3],2]
2021-11-01T04:22:29.7338640Z 150: itk.Image[itk.Vector[itk.F,4],2]
2021-11-01T04:22:29.7396520Z 150: itk.Image[itk.CovariantVector[itk.F,4],2]
2021-11-01T04:22:29.7436410Z 150: itk.Image[itk.F,3]
2021-11-01T04:22:29.7439990Z 150: itk.Image[itk.SymmetricSecondRankTensor[itk.F,3],3]
2021-11-01T04:22:29.7498000Z 150: itk.Image[itk.Vector[itk.F,2],3]
2021-11-01T04:22:29.7537830Z 150: itk.Image[itk.CovariantVector[itk.F,2],3]
2021-11-01T04:22:29.7541550Z 150: itk.Image[itk.Vector[itk.F,3],3]
2021-11-01T04:22:29.7599130Z 150: itk.Image[itk.CovariantVector[itk.F,3],3]
2021-11-01T04:22:29.7639200Z 150: itk.Image[itk.Vector[itk.F,4],3]
2021-11-01T04:22:29.7642800Z 150: itk.Image[itk.CovariantVector[itk.F,4],3]
2021-11-01T04:22:29.7701950Z 150: itk.Image[itk.D,2]
2021-11-01T04:22:29.7740640Z 150: itk.Image[itk.SymmetricSecondRankTensor[itk.D,2],2]
2021-11-01T04:22:29.7802860Z 150: itk.Image[itk.Vector[itk.D,2],2]
2021-11-01T04:22:29.7842220Z 150: itk.Image[itk.CovariantVector[itk.D,2],2]
2021-11-01T04:22:29.7844230Z 150: itk.Image[itk.Vector[itk.D,3],2]
2021-11-01T04:22:29.7904210Z 150: itk.Image[itk.CovariantVector[itk.D,3],2]
2021-11-01T04:22:29.7943630Z 150: itk.Image[itk.Vector[itk.D,4],2]
2021-11-01T04:22:29.7945530Z 150: itk.Image[itk.CovariantVector[itk.D,4],2]
2021-11-01T04:22:29.8005560Z 150: itk.Image[itk.D,3]
2021-11-01T04:22:29.8044950Z 150: itk.Image[itk.SymmetricSecondRankTensor[itk.D,3],3]
2021-11-01T04:22:29.8046770Z 150: itk.Image[itk.Vector[itk.D,2],3]
2021-11-01T04:22:29.8106810Z 150: itk.Image[itk.CovariantVector[itk.D,2],3]
2021-11-01T04:22:29.8146350Z 150: itk.Image[itk.Vector[itk.D,3],3]
2021-11-01T04:22:29.8147830Z 150: itk.Image[itk.CovariantVector[itk.D,3],3]
2021-11-01T04:22:29.8208360Z 150: itk.Image[itk.Vector[itk.D,4],3]
2021-11-01T04:22:29.8249980Z 150: itk.Image[itk.CovariantVector[itk.D,4],3]
2021-11-01T04:22:29.8315330Z 150: itk.Image[itk.UI,2]
2021-11-01T04:22:29.8358090Z 150: itk.Image[itk.UI,3]
2021-11-01T04:22:29.8397900Z 150: itk.Image[itk.UL,2]
2021-11-01T04:22:29.8452810Z 150: itk.Image[itk.UL,3]
2021-11-01T04:22:29.8507560Z 150: itk.Image[itk.ULL,2]
2021-11-01T04:22:29.8537860Z 150: itk.Image[itk.ULL,3]
2021-11-01T04:22:29.8591970Z 150: itk.Image[itk.SI,2]
2021-11-01T04:22:29.8611040Z 150: itk.Image[itk.SI,3]
2021-11-01T04:22:29.8639990Z 150: itk.VectorImage[itk.SS,2]
2021-11-01T04:22:29.8693050Z 150: itk.VectorImage[itk.UC,2]
2021-11-01T04:22:29.8712380Z 150: itk.VectorImage[itk.F,2]
2021-11-01T04:22:29.8741220Z 150: itk.VectorImage[itk.SS,3]
2021-11-01T04:22:29.8794420Z 150: itk.VectorImage[itk.UC,3]
2021-11-01T04:22:29.8813630Z 150: itk.VectorImage[itk.F,3]

```

</details>

thewtex commented on Wrapping/Generators/Python/itk/support/extras.py:
```diff
@@ -801,6 +805,49 @@ def image_from_vtk_image(vtk_image: "vtk.vtkImageData") -> "itkt.ImageBase":
+    """Deserialize an dictionary representing an itk.Image object."""
+    import itk
+
+    ImageType = image_type_from_wasm_type(image_dict['imageType'])
```
The return here is `(ImageType, dtype)` until `image_type_from_wasm_type` is updated to only return `ImageType`

GenevieveBuckley replied:
Ah yes, I should have remembered that. Thanks!

thewtex commented on Wrapping/Generators/Python/itk/support/helpers.py:
```diff
@@ -173,3 +176,130 @@ def image_filter_wrapper(*args, **kwargs):
+    if pixelType not in (1, 2, 3, 10):
+        prefix += str(dimension)
+    prefix += str(dimension)
+    return getattr(itk.Image, prefix), dtype
```
`return getattr(itk.Image, prefix), dtype` -> `getattr(itk.Image, prefix)`

thewtex reviewed (commented):
The

thewtex commented:
> ITK doesn't seem to recognize the array type, I'm getting messages to say it's not wrapped.

A few comments inline that should address this.

GenevieveBuckley commented:
Thanks for helping push this along Matt, I really appreciate it

dzenanz commented:
Maybe the Mac test failure is due to different compiler defaults. Does the test fail with clang on Linux?

thewtex commented:
@GenevieveBuckley thank you so much for your persistence and efforts! :climbing: 

The error on mac does look odd. I will test locally and follow-up.

GenevieveBuckley commented:
> The error on mac does look odd. I will test locally and follow-up.

Thanks Matt!

thewtex commented on Wrapping/Generators/Python/itk/support/extras.py:
```diff
@@ -801,6 +805,49 @@ def image_from_vtk_image(vtk_image: "vtk.vtkImageData") -> "itkt.ImageBase":
+    pixel_arr = itk.array_view_from_image(image)
+    dimension = image.GetImageDimension()
+    componentType, pixelType = image_to_type(image)
+    if 'int64' in componentType:
```
We can remove this work around since this is used with WebAssembly and WebAssembly supports 64-bit integer arrays.

thewtex commented on Wrapping/Generators/Python/Tests/extras.py:
```diff
@@ -162,6 +163,18 @@ def custom_callback(name, progress):
 assert type(image) == itk.Image[itk.RGBPixel[itk.UC], 2]
 
+# Test serialization with pickle
+array = np.random.randint(0, 256, (8, 12))
```
To keep the testing CI build quicker, not all types are wrapped, so we need to add `.astype(np.uint8)`,

```
array = np.random.randint(0, 256, (8, 12)).astype(np.uint8)
```

thewtex commented on Wrapping/Generators/Python/itk/support/helpers.py:
```diff
@@ -173,3 +176,117 @@ def image_filter_wrapper(*args, **kwargs):
     return image_filter_wrapper
+
+
+def image_to_type(itkimage):  # noqa: C901
```
To be more specific and consistent, we could call this `wasm_type_from_image` and return

```
    imageType = dict(
      dimension=dimension,
     componentType=componentType,
     pixelType=pixelType,
     components=itkimage.GetNumberOfComponentsPerPixel()
   )
  return imageType
```

GenevieveBuckley replied:
`dimension` and `componentType` are undefined in that function, are they named something else there instead?

I maaaybe guess:
- ``componentType` == `_python_to_js[mangle]`
- `dimension` == `pixelType`

...but it's late here, so I'm going to stop for today.

GenevieveBuckley replied:
Wait no, I think that's wrong.

...I'll just change the function name, for now. We can talk about this bit more later.

thewtex replied:
`componentType==_python_to_js[mangle]`

Correct

```
dimension = itkimage.GetImageDimension()
components = itkimage.GetNumberOfComponentsPerPixel()
```

thewtex reviewed (commented):
A few minor issues noted inline after testing locally, a squash, and we will be good to rock :rocket: :man_singer:

thewtex commented:
*ITK.macOS.Python* error:

```
  File "/Users/runner/work/1/s/Wrapping/Generators/Python/Tests/extras.py", line 176, in <module>
    assert (pickle.loads(pickle.dumps(image)) == image).all()
AttributeError: 'bool' object has no attribute 'all'
```

Can be addressed by changing

```
    assert (pickle.loads(pickle.dumps(image)) == image).all()
```

to:

```
serialize_deserialize = pickle.loads(pickle.dumps(image))
# verify_input_information checks origin, spacing, direction consistency
comparison = itk.comparison_image_filter(image, serialize_deserialize, verify_input_information=True)
assert np.sum(comparison) == 0.0  
```

thewtex commented:
/azp run ITK.macOS.Python

thewtex commented:
:green_heart:

GenevieveBuckley commented:
Finally the tests are all passing! :tada: 
I've updated the PR title, and squashed all the commits down to one (which has just re-triggered the CI, so I sure hope nothing got lost in that process). With a bit of luck, it should be ready for approval soon.

dzenanz commented on Wrapping/Generators/Python/itk/support/helpers.py:
```diff
@@ -173,3 +176,123 @@ def image_filter_wrapper(*args, **kwargs):
+        mangle = itk.template(component)[1][0]
+        pixelType = 8
+    else:
+        raise RuntimeError('Unrecognized component type: {0}'.format(str(component)))
```
If there is no support for `long` and `long long` types, should we maybe create a more customized error message in case of those?

GenevieveBuckley replied:
Comments below suggest that we can just add support for `long` and `long long` types, so no need to change the error message here.

dzenanz commented on Wrapping/Generators/Python/itk/support/helpers.py:
```diff
@@ -173,3 +176,123 @@ def image_filter_wrapper(*args, **kwargs):
+            return 'int32_t', 1,
+        else:
+            return 'int64_t', 1,
+    if component in (itk.SC, itk.UC, itk.SS, itk.US, itk.SI, itk.UI, itk.F,
```
Are `long` and `long long` types absent from this list on purpose?

GenevieveBuckley replied:
It looks like this list of types was first introduced by @thewtex in this commit [ENH: Use strong component typing _image_to_type](https://github.com/InsightSoftwareConsortium/itkwidgets/commit/88c0e659eca065c46a0eb5e4700d481dca398f78), which was part of this PR [InsightSoftwareConsortium/itkwidgets #323 Unsigned long pixel type](https://github.com/InsightSoftwareConsortium/itkwidgets/pull/323). 

@thewtex can you tell us if we should be adding `long` and `long long` types in this context?

thewtex replied:
Yes, can we please add `long` and `long long` type support by adding `itk.L`, `itk.LL`, `itk.UL`, `itk.ULL` to this list? These were originally not added because JavaScript did not have support, but that is historical now.

GenevieveBuckley replied:
Pretty sure that it needs to be `itk.SL`, `itk.SLL`, `itk.UL`, `itk.ULL`.

When I start ipython and look for `itk.L`/`itk.LL` it tells me "module 'itk' has no attribute 'LL'." Specifying that it's signed or unsigned with `S` or `U` seems to fix that problem.

dzenanz commented on Wrapping/Generators/Python/itk/support/helpers.py:
```diff
@@ -173,3 +176,123 @@ def image_filter_wrapper(*args, **kwargs):
+        else:
+            return 'L'
+    prefix = _pixelType_to_prefix[pixelType]
+    _js_to_python = {
```
`int64_t` is absent from `_python_to_js`. Is this on purpose?

GenevieveBuckley replied:
I'm not sure. Following Matt's suggestion [here](https://github.com/InsightSoftwareConsortium/ITK/issues/1091#issuecomment-863227855), the helper functions are basically copies of [these functions](https://github.com/InsightSoftwareConsortium/itkwidgets/blob/c4f3b158719bbb2720ef2fadcc3df8990a3feb95/itkwidgets/trait_types.py#L175-L270) from itkwidgets. Perhaps @thewtex can provide more detail? It looks like he is the original author of the `_type_to_image`/`_image_to_type` functions.

> To get the state, here is an example that represents the itk.Image state as a Python Dictionary of fundamental types + a np.ndarray (which has pickle support): https://github.com/InsightSoftwareConsortium/itkwidgets/blob/c4f3b158719bbb2720ef2fadcc3df8990a3feb95/itkwidgets/trait_types.py#L163-L172.

thewtex replied:
Yes, these could please be added, thanks!

GenevieveBuckley replied:
Ok, I've added these to `_python_to_js`. I think they should all map to the 64 bit type because all the numpy types are 64 bit when I look at `itk.SLL.dtype` / itk.SL.dtype / etc.

```python
    _python_to_js = {
        ....
        # new code added
        itk.SL: 'int64_t',
        itk.UL: 'uint64_t',
        itk.SLL: 'int64_t',
        itk.ULL: 'uint64_t',
    }
```

thewtex replied:
Yes, we should correct this for Windows, -- `signed long` (`SL`) and `unsigned long` (`UL`) are 32-bit on Windows :man_shrugging: .

GenevieveBuckley replied:
Thank you!

thewtex commented on Wrapping/Generators/Python/itk/support/helpers.py:
```diff
@@ -173,3 +176,123 @@ def image_filter_wrapper(*args, **kwargs):
+        itk.SS: 'int16_t',
+        itk.US: 'uint16_t',
+        itk.SI: 'int32_t',
+        itk.UI: 'uint32_t',
```
We could add the `long`, `long long` support here as in `_long_type()` in `image_type_from_wasm_type`.

GenevieveBuckley replied:
This comment suggests that maybe [this](https://github.com/InsightSoftwareConsortium/ITK/pull/2829#discussion_r746316216) was not quite what you had in mind? 

The `_long_type()` function behaves differently on posix vs windows machines, do we need to treat them differently here too? I'm sorry, I don't have a Windows machine handy, or I'd be able to test this out for myself.

dzenanz replied:
Yes, on Visual Studio and 64-bit Windows, long is still 32-bits. GCC and clang have long as 64-bits on 64-bit machines, so we do need to treat long differently. I saw an `if os.nt` as a distinguisher already used in the code to treat long differently.

thewtex replied:
This size ambiguity of `long` is a reason why we use the explicit type names in the WebAssembly serialization interface. Recently NumPy has moved to prefer more explicit use of primative type sizes, e.g. `np.float64`.

dzenanz commented on Wrapping/Generators/Python/itk/support/helpers.py:
```diff
@@ -173,3 +176,127 @@ def image_filter_wrapper(*args, **kwargs):
+        else:
+            return itk.Image[itk.complex, itk.D], np.float64
+
+    def _long_type():
```
@GenevieveBuckley you could use this or something similar to treat `long` different depending on platform.

GenevieveBuckley replied:
Thank you, I had seen this. My question was more about what result should be returned for different platforms, eg:

I don't have access to a Windows machine, so I had to guess these:
* nt, signed long  -> ?? (I guessed `int64_t`, but from the discussion about it sounds like it should be `int32_t`)
* nt, signed long long -> ?? (I guessed `int64_t`, and I don't see anyone disagreeing)
* nt, unsigned long -> ?? (I guessed `uint64_t`, but from the discussion about it sounds like it should be `uint32_t`)
* nt, unsigned long long -> ?? (I guessed `uint64_t`, and I don't see anyone disagreeing)

I know the right answers on posix, because I can check these on my machine.
* posix, signed long -> should be `int64_t`
* posix, signed long long -> should be `int64_t`
* posix, unsigned long -> should be `uint64_t`
* posix, unsigned long long -> should be `uint64_t`

thewtex reviewed (approved):
Woohoo!! @GenevieveBuckley you rock! :100: :champagne: :sparkler:

GenevieveBuckley commented:
Hooray! :smile: 

It would be good to do another [Dask & ITK blogpost](https://github.com/dask/dask-blog/issues/27), to let people know, so they can try it out. Would you be interested in that?

I'm not familiar with ITK's release schedule, but this looks most likely to be included from the next release candidate (version 5.3rc2 and above).

thewtex commented:
> It would be good to do another Dask & ITK blogpost, to let people know, so they can try it out. Would you be interested in that?

Yes, absolutely!

Through this process, an interesting question came up about zstandard compression with Dask distributed. Since we are leveraging the NumPy array serialization, I am wondering if we could monkey-patch the NumPy serialization, which could help performance beyond ITK (or not)? :thought_balloon: :thinking: 

> I'm not familiar with ITK's release schedule, but this looks most likely to be included from the next release candidate (version 5.3rc2 and above).

Version 5.3rc2 is coming out now, but this will be available in 5.3rc3, coming out in the next week or two. We can test before the 5.3.0 release.

jakirkham commented:
FWIW [Distributed does its own compression]( https://github.com/dask/distributed/blob/7a3ea4c34ae0af6af1137fe3e3bda927aa9eac26/distributed/protocol/compression.py ), which could use any of several compressors (including Zstandard). So this may already do what you want without needing to add this to pickle.

GenevieveBuckley commented:
> Version 5.3rc2 is coming out now, but this will be available in 5.3rc3, coming out in the next week or two. We can test before the 5.3.0 release

Ok, good. John suggests we should re-run the [code from the blogpost](https://blog.dask.org/2019/08/09/image-itk), which should tell us if [this](https://github.com/InsightSoftwareConsortium/ITK/issues/1134) is still a problem or not. (I'm mid-way through attempting to build & install itk locally, but not completely confident I'll be able to do that well. So I can either wait for the v5.3rc03 pre-release, or someone with a local itk development version could do that sooner.)

GenevieveBuckley commented:
And because @thewtex doesn't seem to be on the Dask slack, I'll share a copy of this relevant comment by @jakirkham:
> As to any other remaining pieces, would trust that blogpost more than my own memory at this point
>
> We might want to recheck this GIL issue ( https://github.com/InsightSoftwareConsortium/ITK/issues/1134 ). There is no repro in the issue, but maybe just rerunning the code in the original blog will identify it
>
> This issue ( https://github.com/InsightSoftwareConsortium/ITK/issues/1151 ) is also related to the blog. Basically let’s us drop map_blocks and just use ITK on Dask directly. Though it involves changes in NumPy (like in C or Cython). So Idk if that’s something worth tackling or not ( https://github.com/numpy/numpy/issues/14020 ). It would be very useful, but it could also get into the weeds a bit
>
> So to summarize maybe check the GIL issue and then it should be good for a new blogpost (ideally with simpler code than we had before)

Blogpost link: https://blog.dask.org/2019/08/09/image-itk

jakirkham commented:
To clarify this:

> > As to any other remaining pieces, would trust that blogpost more than my own memory at this point
> > We might want to recheck this GIL issue ( #1134 ). There is no repro in the issue, but maybe just rerunning the code in the original blog will identify it

I think Matt R. & I were seeing lock-ups when trying to run things in parallel with ITK and Dask. So the question would be when running this code today with a recent ITK version do they still occur? Matt M. suggests it may be fixed ( https://github.com/InsightSoftwareConsortium/ITK/issues/1134#issuecomment-862896832 ). Either confirming the issue is fixed or still present would be useful. Including a repro if still present would also be helpful.
